src/accumulate-command-stack.ts

```diff
@@ -1,53 +1,19 @@
-import { Command, CommandStack } from './types';
-import { LEAF, BRANCH } from './constants';
+import { Command } from './types';
+import { LEAF } from './constants';
 
-export function accumulateCommandStack(
-  rootCommand: Command,
-  commandNameAndArgsArgv: string[],
-) {
-  const commandStack: CommandStack = {
-    branches: [],
-    leaf: undefined,
-  };
-
-  function addToCommandStack(command: Command) {
-    switch (command._type) {
-      case BRANCH:
-        commandStack.branches.push(command);
-        break;
-      case LEAF:
-        commandStack.leaf = command;
-        break;
-      default:
-        throw new Error('Unexpected command type');
-    }
+export function accumulateCommandStack(command: Command, argv: string[]): string[] {
+  if (command._type === LEAF) {
+    return argv;
   }
 
-  addToCommandStack(rootCommand);
+  const found = command.subcommands.find(subcommand => subcommand.name === argv[0]);
 
-  let badCommandName: string | undefined = undefined;
-  let index = 0;
-  while (
-    typeof badCommandName === 'undefined' &&
-    !commandStack.leaf &&
-    index < commandNameAndArgsArgv.length
-  ) {
-    const maybeCommandName = commandNameAndArgsArgv[index];
-    index = index + 1;
-    const branch = commandStack.branches.slice(-1)[0];
-    const nextCommand = branch.subcommands.find(
-      subcommand => subcommand.name === maybeCommandName,
-    );
-    if (!nextCommand) {
-      badCommandName = maybeCommandName;
-      break;
-    }
-    addToCommandStack(nextCommand);
+  if (!found) {
+    command.next = undefined;
+    return argv;
   }
-  const argsArgv = commandNameAndArgsArgv.slice(index);
-  return {
-    commandStack,
-    argsArgv,
-    badCommandName,
-  };
+
+  command.next = found;
+
+  return accumulateCommandStack(found, argv.slice(1));
```

Comment:
^^ Calls accumulateCommandStack recursively. Notice how much shorter this function is now. 50 lines down to 18!